### ISSUE TITLE
Riverscapes Context Metrics

### ISSUE THREAD

**Issue by philipbaileynar** [Riverscapes/riverscapes-tools#574]
Each riverscapes tool is going to start producing "metrics", starting with Riverscapes Context. The goal is to produce a simple dataset that summarizes the contents of the project that can then be scraped by other systems:

- the metrics can appear in the HTML report for each tool run.
- the metrics can be summarized on the data warehouse welcome screen (e.g. "Warehouse contains X million miles of river")
- the metrics can be aggregated into the network attribute tool (an API for querying riverscape metrics for a lat/long).

@MattReimer and I are still discussing where the metrics will be stored. Our current thinking is that they will be written to a file called `metrics.json` at the root of each project and references in the `project.rs.xml`. We used a similar pattern for storing the spatial extent of each project in a JSON file.

For riverscapes context the metrics will fall into the following categories:

# Configuration

- All the following metrics will be reported at two scales:
    - The entire project
    - Each individual HUC12
- Units - All metrics will be calculated and stored as metric values. Software that displays the values will be responsible for converting them to other required units (e.g. river miles).
- Which datasets and which attributes of each dataset that are to be processed should be abstracted to a dictionary at the top of a Python code file somewhere. This should enable quick and easy configuration of which datasets are processed.

# Floating Point Rasters

Loop over all the floating point rasters in the project (e.g. PRISM, DEM, slope) and report the following metrics. Note that the [raster buffer stats](https://github.com/Riverscapes/riverscapes-tools/blob/9d885ce03a121b55767b00abdd64747c1e1d4384/lib/commons/rscommons/raster_buffer_stats.py#L119-L157) already does all the calculations necessary!!!!!!

- cell area
- minimum value
- maximum value
- average value
- sum of values cells
- count of data (i.e. non-NoData) cells

(not the sum and count metrics allow for averages to be computed at other scales.)

# Integer Rasters

Loop over all the categorical rasters in the project (e.g. LandFire) and report the following:

- count of cells for each category
- area of cells

# Polygon Datasets

- area of each attribute specified in the configuration dictionary
- percent of the processing area (either project or HUC12) that the area represents.

# Polyline Datasets

- length of each attribute specified in the configuration dictionary
- percent of the total length (either project or HUC 12) that the length represents.

# Point Datasets

- Count of features by attribute specified in configuration dictionary

# Example JSON

Needs confirmation before proceeding:

```json
{
  "project": {
    "name": "My HUC8 Name",
    "area": 12345.4556,
    "metrics": {
      "rasters": {
        "floating_point": {
          "dem": {
            "cell_size": 0.1,
            "max": 1234.567,
            "min": 0.00,
            "avg": 500.00,
            "count": 1234567,
            "sum": 123456.789
          },
          "slope": {
            "cell_size": 0.1,
            "max": 1234.567,
            "min": 0.00,
            "avg": 500.00,
            "count": 1234567,
            "sum": 123456.789
          }
        },
        "categorical": {
          "landfire_evt": {
            "cell_size": 0.3,
            "categories": [
              {
                "ponderosa": {
                  "area": 123456.8,
                  "count": 1234567
                }
              }
            ]
          },
          "landfire_bps": {
            "cell_size": 0.3,
            "categories": [
              {
                "ponderosa": {
                  "area": 123456.8,
                  "count": 1234567
                }
              }
            ]
          }
        }
      },
      "vector": {
        "polygon": {
          "ecoregions3": {
            "name": {
              "dry scrubby desert": {
                "area": 123456,
                "proportion": 0.345
              },
              "wetland": {
                "area": 123456,
                "proportion": 0.345
              }
            }
          }
        }
      }
    }
  },
  "huc12": {
    "1704123453": {
      "name": "My HUC12 Name 1",
      "area": 12345,
      "metrics": {
        // same structure as project metrics above
      }
    },
    "1704123454": {
      "name": "My HUC12 Name 2",
      "area": 67890,
      "metrics": {
        // same structure as project metrics above
      }
    }
  }
}
```

**Comment by joewheaton** [comment]
I like this @philipbaileynar. For floating point rasters, a question: is it possible to provide summary statistics for rasters based on the legends specified in the QML for that raster tied to the business logic? The basic stats (max, min, range, average, etc.) are nice, but what you really tend to care about are the proportions in the categories specified by our QML (e.g. slope analysis categories).

**Comment by MattReimer** [comment]
@philipbaileynar and I talked a bit about this yesterday. None if it seems like a problem but I want to make sure we do it in a way that is extensible and doesn't need a complete re-run when we add a metric or fix a bug with them. 

We had talked about pre-calculating some of the easy stats during RSContext creation. things like area, min, max avg etc; stats that are generic, useful and uncontroversial. Calculating these for an RSContext project from huc8 to huc12 is not a big deal in my opinion. 

The biggest challenges from my point of view are:

1. Keep the `riverscapes-tools` repo clean and clear from speculative, iterative commits to add/remove/fix/calculate metrics.
2. Keep the cost of adding new metrics or re-running metric calcs as low as humanly possible. 
3. Keep the "spidery-ness" of this solution down so we don't need 11 tools and 37 steps to perform things.
4. Make the metric code as modular as possible. Adding a new metric should be as easy as copying a python file and filling in a few lines of python. Most of the repetitive stuff should be genericized away into `rscommons` like:
    * All the iteration over projects, 
    * writing to JSON, XML reading etc
    * windowed looping over raster cell values 
    * iterating over features in a vector.

We can investigate using cybercastor to run metric calc/re-calc. We'll start manual, of course but this is a perfect case where cybercastor could run some generic python code and API calls for us

Then we would have a python tool that would consume `riverscapes-tools/lib/common`. This tool would:

1. Loop over projects in the warehouse
2. Collect any bounds or relevant base stats from RSContext
3. Calculate any of our advanced metrics
4. Write them back to the Network Attribute API. 
5. (optional) we could also write these stats back to the warehouse if we could agree on where they should go

> is it possible to provide summary statistics for rasters based on the legends specified in the QML for that raster tied to the business logic?

@joewheaton I think the QML would be hard to parse but luckily we have JSON for webrave that should be doing the exact same work. We should talk about how/when/where this gets calculated. There is a bigger problem here that there is not a 1:1 relationship between rasters and symbologies so choosing which symbology to use for a given raster can be challenging. 

I look forward to a bigger discussion here.

**Comment by MattReimer** [comment]
I think this issue is done and superseded by more specific and granular tickets like

* #599

And all this folds nicely into the new https://github.com/Riverscapes/watershed-attribute-api

re-open if you disagree
